### PR TITLE
[Coop] Convert System.Globalization.{CalendarData,CultureData,CultureInfo,RegionInfo} icalls.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build  failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/metadata/object.c [hunk 1]
```diff
@@ -5858,17 +5871,16 @@ mono_object_new_alloc_specific_checked (MonoVTable *vtable, MonoError *error)
 	return object_new_common_tail (o, vtable->klass, error);
 }
 
-MonoObjectHandle
-mono_object_new_alloc_by_vtable (MonoVTable *vtable, MonoError *error)
+void
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/metadata/object.c]: Changed to this in latest. And then constant not-writable `NULL_HANDLE` or writable and can fail under low memory `mono_new_null()`?

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build Linux Aarch64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
1. This needs a new monolite since rebasing to master.
2. Should this be split into smaller? For example, along leaf namespace? Or do all the interface changes first?

**Comment by jaykrell** [comment]
Needs merge and monolite.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

File: mono/metadata/object-internals.h [hunk 2]
```diff
@@ -1673,13 +1687,14 @@ mono_string_handle_length (MonoStringHandle s);
 
 #endif
 
+#define mono_string_chars_unsafe(s) (mono_string_chars_internal (MONO_HANDLE_RAW (s)))
```
Comment (jaykrell) [inline mono/metadata/object-internals.h]: Again should avoid this.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build Linux WebAssembly

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 C++

**Comment by lambdageek** [comment]
I don't think this is the best approach.

We should try to do more from managed.  Maybe make some marshalling structs that we can fill from native and then convert to managed objects in managed.  Here's how corefx does it, for example https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Globalization/CalendarData.Unix.cs and https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/Interop/Unix/System.Globalization.Native/Interop.Calendar.cs  (To be honest the specifics aren't great - all those native callbacks seem too general - but putting more of the initialization logic in managed is a good idea).

I'm still not a fan of the `_assign`-style handle functions.  I don't think it's a pattern we should push.

I think the `utf8` and `utf8z` functions pick the wrong thing as the default (ie shorter name): most of our code assumed nul-terminated utf8 data in a `const char*`; the bytes+length form should have the longer name.

**Comment by jaykrell** [comment]
The default naming is for performance. Strlen is slow and should be avoided and should be made more visible. So it gets the longer name.

**Comment by jaykrell** [comment]
The existing structure seemed straightforward enough to work with.
Corefx is just going right to the Win32 API, or a Unix PAL implementation thereof.
I can try to look into restructuring it instead some time..

File: mono/metadata/object.c [hunk 3]
```diff
@@ -7589,12 +7652,21 @@ mono_string_to_utf8_checked_internal (MonoString *s, MonoError *error)
...
 	return mono_utf16_to_utf8 (mono_string_chars_internal (s), s->length, error);
 }
 
+char *
+mono_string_to_utf8len (MonoStringHandle s, gsize *utf8len, MonoError *error)
+{
+	*utf8len = 0;
+	if (MONO_HANDLE_IS_NULL (s))
+		return NULL;
+	guint gchandle = 0;
```
Comment (jaykrell) [inline mono/metadata/object.c]: This should use the safepoints macros instead of `gchandle`.

**Pull request closed without merging**